File: src/main/java/com/lessmarkup/userinterface/model/user/RegisterModel.java
package com.lessmarkup.userinterface.model.user;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.exceptions.CommonException;
import com.lessmarkup.interfaces.recordmodel.InputField;
import com.lessmarkup.interfaces.recordmodel.InputFieldType;
import com.lessmarkup.interfaces.recordmodel.RecordModel;
import com.lessmarkup.interfaces.recordmodel.RecordModelCache;
import com.lessmarkup.interfaces.recordmodel.RecordModelDefinition;
import com.lessmarkup.interfaces.security.UserSecurity;
import com.lessmarkup.interfaces.system.SiteConfiguration;

public class RegisterModel extends RecordModel<RegisterModel> {

    private final DataCache dataCache;
    private final UserSecurity userSecurity;

    private String email;
    private String name;
    private boolean generatePassword;
    private String password;
    private boolean showUserAgreement;
    private String userAgreement;
    private boolean agree;

    @Inject
    public RegisterModel(DataCache dataCache, UserSecurity userSecurity) {
        this.dataCache = dataCache;
        this.userSecurity = userSecurity;
    }

    public JsonElement getRegisterObject() {
        SiteConfiguration siteProperties = dataCache.get(SiteConfiguration.class);

        if (!siteProperties.getHasUsers()) {
            throw new CommonException(LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.CANNOT_REGISTER_NEW_USER));
        }

        userAgreement = siteProperties.getUserAgreement();
        showUserAgreement = !StringHelper.isNullOrWhitespace(siteProperties.getUserAgreement());

        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);

        JsonObject ret = new JsonObject();
        ret.add("registerObject", JsonSerializer.serializePojoToTree(this));
        ret.addProperty("modelId", modelCache.getDefinition(RegisterModel.class).get().getId());

        return ret;
    }

    public JsonElement register() {
        SiteConfiguration siteProperties = dataCache.get(SiteConfiguration.class);

        if (!siteProperties.getHasUsers()) {
            throw new CommonException(LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.CANNOT_REGISTER_NEW_USER));
        }

        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);
        RecordModelDefinition definition = modelCache.getDefinition(RegisterModel.class).get();
        definition.validateInput(JsonSerializer.serializePojoToTree(this), true);

        userSecurity.createUser(name, password, email, false, false);

        boolean loggedIn = RequestContextHolder.getContext().getCurrentUser().loginWithPassword(email, password, false, false, true, null);

        JsonObject ret = new JsonObject();

        ret.addProperty("userName", name);
        ret.addProperty("showConfiguration", false);
        ret.addProperty("userLoggedIn", loggedIn);

        return ret;
    }

    @InputField(type = InputFieldType.EMAIL, textId = TextIds.EMAIL, required = true)
    public void setEmail(String email) { this.email = email; }
    public String getEmail() { return email; }

    @InputField(type = InputFieldType.TEXT, textId = TextIds.USER_NAME, required = true)
    public void setName(String name) { this.name = name; }
    public String getName() { return name; }

    @InputField(type = InputFieldType.CHECK_BOX, textId = TextIds.GENERATE_PASSWORD)
    public void setGeneratePassword(boolean generatePassword) { this.generatePassword = generatePassword; }
    public boolean isGeneratePassword() { return generatePassword; }

    @InputField(type = InputFieldType.PASSWORD_REPEAT, textId = TextIds.PASSWORD, required = true, visibleCondition = "!generatePassword")
    public void setPassword(String password) { this.password = password; }
    public String getPassword() { return password; }

    public void setShowUserAgreement(boolean showUserAgreement) { this.showUserAgreement = showUserAgreement; }
    public boolean isShowUserAgreement() { return showUserAgreement; }

    @InputField(type = InputFieldType.RICH_TEXT, textId = TextIds.USER_AGREEMENT, visibleCondition = "showUserAgreement")
    public void setUserAgreement(String userAgreement) { this.userAgreement = userAgreement; }
    public String getUserAgreement() { return userAgreement; }

    @InputField(type = InputFieldType.CHECK_BOX, textId = TextIds.AGREE, required = true, visibleCondition = "showUserAgreement")
    public void setAgree(boolean agree) { this.agree = agree; }
    public boolean isAgree() { return agree; }
}

File: src/main/scala/com/lessmarkup/userinterface/model/user/RegisterModel.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.model.user

import com.google.gson.{JsonElement, JsonObject}
import com.google.inject.Inject
import com.lessmarkup.{Constants, TextIds}
import com.lessmarkup.framework.helpers.{JsonSerializer, LanguageHelper, StringHelper}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.{InputField, InputFieldType}
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.exceptions.CommonException
import com.lessmarkup.interfaces.recordmodel.{RecordModel, RecordModelCache, RecordModelDefinition}
import com.lessmarkup.interfaces.security.UserSecurity
import com.lessmarkup.interfaces.system.SiteConfiguration

class RegisterModel @Inject() (dataCache: DataCache, userSecurity: UserSecurity) extends RecordModel[RegisterModel] {
  private var email: String = null
  private var name: String = null
  private var generatePassword: Boolean = false
  private var password: String = null
  private var showUserAgreement: Boolean = false
  private var userAgreement: String = null
  private var agree: Boolean = false

  def getRegisterObject: JsonElement = {
    val siteProperties: SiteConfiguration = dataCache.get(classOf[SiteConfiguration])
    if (!siteProperties.hasUsers) {
      throw new CommonException(LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.CANNOT_REGISTER_NEW_USER))
    }
    userAgreement = siteProperties.userAgreement
    showUserAgreement = !StringHelper.isNullOrWhitespace(siteProperties.userAgreement)
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val ret: JsonObject = new JsonObject
    ret.add("registerObject", JsonSerializer.serializePojoToTree(this))
    ret.addProperty("modelId", modelCache.getDefinition(classOf[RegisterModel]).get.getId)
    ret
  }

  def register: JsonElement = {
    val siteProperties: SiteConfiguration = dataCache.get(classOf[SiteConfiguration])
    if (!siteProperties.hasUsers) {
      throw new CommonException(LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.CANNOT_REGISTER_NEW_USER))
    }
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val definition: RecordModelDefinition = modelCache.getDefinition(classOf[RegisterModel]).get
    definition.validateInput(JsonSerializer.serializePojoToTree(this), isNew = true)
    userSecurity.createUser(name, password, email, preApproved = false, generatePassword = false)
    val loggedIn: Boolean = RequestContextHolder.getContext.getCurrentUser.loginWithPassword(email, password, savePassword = false, allowAdmin = false, allowRegular = true, null)
    val ret: JsonObject = new JsonObject
    ret.addProperty("userName", name)
    ret.addProperty("showConfiguration", false)
    ret.addProperty("userLoggedIn", loggedIn)
    ret
  }

  @InputField(fieldType = InputFieldType.EMAIL, textId = TextIds.EMAIL, required = true) def setEmail(email: String) {
    this.email = email
  }

  def getEmail: String = {
    email
  }

  @InputField(fieldType = InputFieldType.TEXT, textId = TextIds.USER_NAME, required = true) def setName(name: String) {
    this.name = name
  }

  def getName: String = {
    name
  }

  @InputField(fieldType = InputFieldType.CHECK_BOX, textId = TextIds.GENERATE_PASSWORD) def setGeneratePassword(generatePassword: Boolean) {
    this.generatePassword = generatePassword
  }

  def isGeneratePassword: Boolean = {
    generatePassword
  }

  @InputField(fieldType = InputFieldType.PASSWORD_REPEAT, textId = TextIds.PASSWORD, required = true, visibleCondition = "!generatePassword") def setPassword(password: String) {
    this.password = password
  }

  def getPassword: String = {
    password
  }

  def setShowUserAgreement(showUserAgreement: Boolean) {
    this.showUserAgreement = showUserAgreement
  }

  def isShowUserAgreement: Boolean = {
    showUserAgreement
  }

  @InputField(fieldType = InputFieldType.RICH_TEXT, textId = TextIds.USER_AGREEMENT, visibleCondition = "showUserAgreement") def setUserAgreement(userAgreement: String) {
    this.userAgreement = userAgreement
  }

  def getUserAgreement: String = {
    userAgreement
  }

  @InputField(fieldType = InputFieldType.CHECK_BOX, textId = TextIds.AGREE, required = true, visibleCondition = "showUserAgreement") def setAgree(agree: Boolean) {
    this.agree = agree
  }

  def isAgree: Boolean = {
    agree
  }
}